torchtune/modules/loss/cross_entropy_loss.py

```diff
@@ -143,3 +143,95 @@ def forward(
             return total_loss
         else:
             return total_loss / total_elements
+
+class LigerFusedCrossEntropyLoss(nn.Module, SFTLoss):
+    """Memory efficient Cross-entropy loss that uses fused CUDA kernels to compute the loss.
+    Combines the linear projection with the cross-entropy calculation for better performance
+    and memory efficiency.
+
+    Linear cross entropy is computed in a single fused operation. You need to skip the final 
+    projection layer in your model and pass it to the loss instead. You can setup the loss 
+    with the model as shown below.
+
+    >>> model = Transformer(...)
+    >>> loss = LigerFusedCrossEntropyLoss(...)
+    >>> loss.set_model_output(model)
+    """
+
+    def __init__(
+        self,
+        ignore_index: int = -100
+    ):
+        """Initialize the LigerFusedCrossEntropyLoss.
+
+        Args:
+            ignore_index (int): Index to ignore in the target tensor. Default is -100.
+        """
+        super().__init__()
+        self.linear_projection = None
+        self.ignore_index = ignore_index
+
+    def set_model_output(self, model: nn.Module) -> None:
+        """Modify model output to match the expected input for the loss function.
+        
+        Args:
+            model (nn.Module): The model whose output layer will be used for the loss computation.
+        """
+        model.skip_output_layer = True
+        self.linear_projection = model.output
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        targets: torch.Tensor,
+    ) -> torch.Tensor:
+        """Compute the fused linear cross entropy loss.
+
+        Args:
+            hidden_states (torch.Tensor): Hidden state of the model, pre projection. Shape ``[bsz, seq_len, emb_dim]``
+            targets (torch.Tensor): Labels for the model. Shape ``[bsz, seq_len]``
+
+        Returns:
+            torch.Tensor: The computed loss value
+
+        Raises:
+            RuntimeError: If set_model_output() was not called before forward
+        """
+        if self.linear_projection is None:
+            raise RuntimeError("Must call set_model_output() before forward()")
+        if isinstance(hidden_states, DTensor):
+            hidden_states = hidden_states.full_tensor()
+        
+        orig_w = self.linear_projection.weight
+        if isinstance(orig_w, DTensor):
+            mesh, placements = orig_w.device_mesh, orig_w.placements
+            w = orig_w.full_tensor().detach().clone().requires_grad_(True)
```

Comment:
Can you tell me more about what you're doing here? Does the liger loss require you to detach the weight? Why detach it only to manually register that gradients get reapplied? Also, I don't think you'd want to register a hook every forward pass.